### PR TITLE
`BatchFeature`: Convert `List[np.ndarray]` to `np.ndarray` before converting to pytorch tensors

### LINKED ISSUE #14307: `BatchFeature` performance improvement: convert `List[np.ndarray]` to `np.ndarray` before converting to pytorch tensors
# 🚀 Feature request
@NielsRogge, @sgugger
When using a `FeatureExtractor` for images and passing `List[np.ndarray]` with `return_tensors="pt"`, the following warning is outputted:
```
.../lib/python3.8/site-packages/transformers/feature_extraction_utils.py:158: 
UserWarning: Creating a tensor from a list of numpy.ndarrays is extremely slow. 
Please consider converting the list to a single numpy.ndarray with numpy.array() before converting to a tensor. 
(Triggered internally at  ../torch/csrc/utils/tensor_new.cpp:201.)
``` 

As reported in https://github.com/pytorch/pytorch/issues/13918, a significant performance improvement can be obtained by using `torch.tensor` on a `numpy.ndarray` instead of on `List[numpy.ndarray]`.

I think a possible solution would be https://github.com/huggingface/transformers/pull/14306:
https://github.com/huggingface/transformers/blob/05fed8bf19547161707aec882e081023378608a7/src/transformers/feature_extraction_utils.py#L136-L144

### PR DESCRIPTION
**Pull request by eladsegal**
# What does this PR do?

Fixes #14307

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by sgugger** [comment]
Thanks for your PR!
I think the check needs to be a bit more thorough: we need to also check all the elements of the list are `np.ndarray` (or at least the first one), according to the warning at least.

**Comment by eladsegal** [comment]
Thanks @sgugger,
I added a check for the first element, but here are my thoughts about it:
Iterating through the full list doesn't seem helpful - what would you do if half of the list is `List` and the other half is `np.array`? It would still be more efficient to call `np.array` in such a case.
It's the same thing for checking just the first element, but at least the check doesn't have the cost of iterating through the full list.
Therefore, I think that it's fine to always pass the list to `np.array` without checking it.

File: src/transformers/feature_extraction_utils.py [hunk 1]
```diff
@@ -138,7 +139,13 @@ def convert_to_tensors(self, tensor_type: Optional[Union[str, TensorType]] = Non
                 raise ImportError("Unable to convert output to PyTorch tensors format, PyTorch is not installed.")
             import torch
 
-            as_tensor = torch.tensor
+            def as_tensor(value):
+                if isinstance(value, Sequence) and len(value) > 0:
+                    first_elem = value[0]
+                    if isinstance(first_elem, np.ndarray):
+                        value = np.array(value)
```
Comment (sgugger) [inline src/transformers/feature_extraction_utils.py:146]:
```suggestion
                if isinstance(value, (list, tuple)) and len(value) > 0 and isinstance(value[0], np.ndarray):
                    value = np.array(value)
```
I'm not super enthusiastic about using `Sequence` here as it's wider than lists/tuples. We check for lists/tuples usually.

**Comment by sgugger** [comment]
Failure is unrelated and already fixed on master.
Thanks again for your work on this PR!

**Pull request merged**